presto-orc/src/main/java/com/facebook/presto/orc/writer/StreamOrderingLayout.java

```diff
@@ -0,0 +1,219 @@
+        ImmutableMap.Builder<ColIdSequenceId, List<StreamDataOutput>> seqIdToStreamBuilder = ImmutableMap.<ColIdSequenceId, List<StreamDataOutput>>builder();
+        Map<Integer, Set<DwrfProto.KeyInfo>> colIdToKeyIds = new HashMap<>();
+
+        for (Map.Entry<Integer, List<DwrfProto.KeyInfo>> orderingConfig : config.getStreamOrdering().entrySet()) {
+            colIdToKeyIds.put(orderingConfig.getKey(), new HashSet<>(orderingConfig.getValue()));
+        }
+
+        // iterate through all the encodings and if the encoding has additional sequence encodings put it in the map
+        for (Map.Entry<Integer, ColumnEncoding> entry : nodeIdToColumnEncodings.entrySet()) {
+            int nodeId = entry.getKey();
+            int colId = nodeIdToColumnId.get(nodeId);
+
+            if (entry.getValue().getAdditionalSequenceEncodings().isPresent() && colIdToKeyIds.containsKey(colId)) {
+                // add entries only if stream ordering contains the column ID
+                for (Map.Entry<Integer, DwrfSequenceEncoding> seqIdToEncoding : entry.getValue().getAdditionalSequenceEncodings().get().entrySet()) {
+                    int sequenceId = seqIdToEncoding.getKey();
+                    DwrfProto.KeyInfo keyInfo = seqIdToEncoding.getValue().getKey();
+                    // add the stream only if it is present in the stream ordering config
+                    if (colIdToKeyIds.get(colId).contains(keyInfo)) {
+                        keyIdToSeqIdBuilder.put(new ColIdKeyId(colId, keyInfo), sequenceId);
+                        seqIdToStreamBuilder.put(new ColIdSequenceId(colId, sequenceId), new ArrayList<>());
+                    }
+                }
+            }
+        }
+        return new StreamMetadata(seqIdToStreamBuilder.build(), keyIdToSeqIdBuilder.build());
+    }
+
+    @Override
+    public void reorder(List<StreamDataOutput> dataStreams)
+    {
+        throw new RuntimeException("StreamOrderingLayout not supported for this API");
+    }
+
+    @Override
+    public void reorder(
+            List<StreamDataOutput> dataStreams,
+            Map<Integer, Integer> nodeIdToColumnId,
+            Map<Integer, ColumnEncoding> nodeIdToColumnEncodings)
+    {
+        List<StreamDataOutput> nonReorderStreams = new ArrayList<>();
+
+        StreamMetadata metadata = getStreamMetadata(nodeIdToColumnId, nodeIdToColumnEncodings, config);
+        Map<ColIdSequenceId, List<StreamDataOutput>> seqIdToStreams = metadata.seqIdToStreams;
+        for (StreamDataOutput dataOutput : dataStreams) {
+            int nodeId = dataOutput.getStream().getColumn();
+            int seqId = dataOutput.getStream().getSequence();
+            int colId = nodeIdToColumnId.get(nodeId);
+            // only if sequence ID > 0, we do a look up in seqIdToStreams
+            if (seqId > 0) {
+                List<StreamDataOutput> streams = seqIdToStreams.get(new ColIdSequenceId(colId, seqId));
```

Comment:
Creating a new `new ColIdSequenceId(colId, seqId)` for the lookup seems like a waste. Should we consider using a long key which is combined of two `column << 32 + seq` ints?